app/models/manageiq/providers/azure/cloud_manager/metrics_capture.rb

```diff
@@ -77,76 +77,135 @@ def perf_collect_metrics(interval_name, start_time = nil, end_time = nil)
+          timestamp = parse_timestamp(metric_value.time_stamp)
+          (memo[timestamp] ||= []) << metric_value.average
+        end
+        [metric.name.value, timestamped_values]
+      end.to_h
 
-      counter_values = api_counters_info.each_with_object({}) do |counter_info, memo|
-        counter_metrics = metrics.values_at(*counter_info.metrics).compact.flatten
+      ## Counters, which are available only through legacy API and storage account (raw)
 
-        # { timestamp => [value, ...], ... }
-        timestamped_values = counter_metrics.each_with_object({}) do |metric_value, ts_memo|
-          timestamp = Time.zone.parse(metric_value.time_stamp)
-          (ts_memo[timestamp] ||= []) << metric_value.average
+      storage_conn = Azure::Armrest::StorageAccountService.new(@connection)
+      storage_accounts = storage_conn.list_all
+
+      metric_definitions, _timings = Benchmark.realtime_block(:capture_counters) do
+        # azure-armrest gem needs to be modified to accept version ('list_definitions' method)
+        response = metrics_conn.send(:rest_get, definitions_url.to_s)
+        Azure::Armrest::ArmrestCollection.create_from_response(response, Azure::Armrest::Insights::MetricDefinition)
+      end
+
+      metric_definitions.each do |metric_definition|
+        metric_name = metric_definition.name.value
+        next unless METRIC_NAMES[:raw].include?(metric_name)
+
+        metric_availability = metric_definition.metric_availabilities.detect { |ma| ma.time_grain == INTERVAL_1_MINUTE }
+        next unless metric_availability
+
+        timestamped_values = metrics[metric_name] ||= {}
+
+        metric_location = metric_availability.location
+
+        storage_account_name = URI.parse(metric_location.table_endpoint).host.split('.').first
+        storage_account      = storage_accounts.find { |account| account.name == storage_account_name }
+        storage_account_keys = storage_conn.list_account_keys(storage_account.name, storage_account.resource_group)
+        storage_account_key  = storage_account_keys.fetch('key1')
+
+        metric_location.table_info.each do |table_info|
+          t_start_time = parse_timestamp(table_info.start_time)
+          t_end_time   = parse_timestamp(table_info.end_time).end_of_day
+          next unless time_interval.overlaps?(t_start_time..t_end_time)
+
+          filter = [
+            "CounterName eq '#{metric_name}'",
+            "PartitionKey eq '#{metric_location.partition_key}'",
+            "Timestamp ge datetime'#{@start_time}'",
+            "Timestamp le datetime'#{@end_time}'",
+          ].join(' and ')
+
+          table_data, _timings = Benchmark.realtime_block(:capture_counters) do
+            storage_account.table_data(
+              table_info.table_name,
+              storage_account_key,
+              :filter => filter,
+              :select => 'TIMESTAMP,Average',
+              :all    => true
+            )
```

Comment:
This is an interesting API, do they basically give you SQL level access to the metrics db over REST?

Reply:
yes. problem is that the table names are unpredictable, and I found only one way to get them so far (same as it was in the previous version of this code file).